mage_output_file.go generated with missing targets

I have a project that is basically a library of mage targets. There are targets defined in several sub-packages and two in the root package. I recently added a magefile so that I could use some of the targets in the library on the library itself, however I occasionally encounter this error when using the magefile:
```
$ go run mage/mage.go -l -debug
DEBUG: 23:39:27.668073 getting all non-mage files in .
DEBUG: 23:39:27.757209 found non-mage files magetest.go
DEBUG: 23:39:27.757232 marked file as non-mage: "magetest.go"
DEBUG: 23:39:27.757235 getting all files plus mage files
DEBUG: 23:39:27.795015 time to scan for Magefiles: 126.974683ms
DEBUG: 23:39:27.795036 found magefiles: magefile.go
DEBUG: 23:39:27.805403 output exe is  /Users/wasche/.magefile/8770b1cb20f3d900f5bebc4498ad7ffa9ba7a519
DEBUG: 23:39:27.816255 go build cache exists, will ignore any compiled binary
DEBUG: 23:39:27.816271 parsing files
DEBUG: 23:39:27.816620 time parse Magefiles: 342.668µs
DEBUG: 23:39:27.816631 found mage:import: github.com/wfscheper/magetest
DEBUG: 23:39:27.816634 found mage:import: github.com/wfscheper/magetest/foo (foo)
DEBUG: 23:39:27.816641 getting import package "github.com/wfscheper/magetest/foo", alias "foo"
DEBUG: 23:39:27.909592 parsing imported package "github.com/wfscheper/magetest/foo" from dir "/Users/wasche/git/magetest/foo"
DEBUG: 23:39:27.909984 found target Foo
DEBUG: 23:39:27.910016 time parse Magefiles: 397.002µs
DEBUG: 23:39:27.910020 setting alias "foo" and package "foo" on func Foo
DEBUG: 23:39:27.997637 parsing imported package "github.com/wfscheper/magetest" from dir "/Users/wasche/git/magetest"
DEBUG: 23:39:27.998022 time parse Magefiles: 356.197µs
DEBUG: 23:39:27.998042 Creating mainfile at mage_output_file.go
DEBUG: 23:39:27.998165 writing new file at mage_output_file.go
DEBUG: 23:39:27.998814 compiling to /Users/wasche/.magefile/8770b1cb20f3d900f5bebc4498ad7ffa9ba7a519
DEBUG: 23:39:27.998820 compiling using gocmd: go
DEBUG: 23:39:28.135084 running go build -o /Users/wasche/.magefile/8770b1cb20f3d900f5bebc4498ad7ffa9ba7a519 magefile.go mage_output_file.go
# command-line-arguments
./mage_output_file.go:19:2: imported and not used: "github.com/wfscheper/magetest" as magetest_mageimport
DEBUG: 23:39:28.304639 time to compile Magefile: 169.424177ms
Error: error compiling magefiles
exit status 1
```

Running with `-keep` reveals that mage detected the `magetest` top-level import, but somehow failed to parse any targets, thus the unused import that causes the compile failure.

I have made a simple git repo that reproduces the issue at https://github.com/wfscheper/magetest. There are two other branches in that repo: working and dot-import. The [working](https://github.com/wfscheper/magetest/tree/working) branch shows the only way I've found to get this to work. The [dot-import](https://github.com/wfscheper/magetest/tree/dot-import) branch is an example of an attempted work-around that still exhibits the same problem.

I'm experiencing the same issue, but it is very intermittent.

I keep running `mage -f` and tried also `mage -f -keep` and can also see that the `mage_output_file.go` does not have generated targets from my package, which is in the same repository.

Unfortunately, it's very random. I've ran it about 20 times and kept getting this error
```
./mage_output_file.go:19:2: imported and not used: "github.com/<user>/<repo> as protobuf_mageimport
Error: error compiling magefiles
```
... but without changing any source code, on 21st attempt it has worked and shown me all the expected targets in the output. Then again about 20 runs were with the above failures and then again one occurrence of a working code.

@natefinch what could possibly cause this non-deterministic compilation?

--- a/parse/parse.go
+++ b/parse/parse.go
@@ -230,6 +230,7 @@ func getNamedImports(gocmd string, pkgs map[string]string) ([]*Import, error) {
 	return imports, nil
 }
 
+// getImport returns the metadata about a package that has been mage:import'ed.
 func getImport(gocmd, importpath, alias string) (*Import, error) {
 	out, err := internal.OutputDebug(gocmd, "list", "-f", "{{.Dir}}||{{.Name}}", importpath)
 	if err != nil {
@@ -241,7 +242,17 @@ func getImport(gocmd, importpath, alias string) (*Import, error) {
 	}
 	dir, name := parts[0], parts[1]
 	debug.Printf("parsing imported package %q from dir %q", importpath, dir)
-	info, err := Package(dir, nil)
+
+	// we use go list to get the list of files, since go/parser doesn't differentiate between
+	// go files with build tags etc, and go list does. This prevents weird problems if you
+	// have more than one package in a folder because of build tags.
+	out, err = internal.OutputDebug(gocmd, "list", "-f", `{{join .GoFiles "||"}}`, importpath)
+	if err != nil {
+		return nil, err
+	}
+	files := strings.Split(out, "||")
+
+	info, err := Package(dir, files)
 	if err != nil {
 		return nil, err
 	}
@@ -253,6 +264,7 @@ func getImport(gocmd, importpath, alias string) (*Import, error) {
 	return &Import{Alias: alias, Name: name, Path: importpath, Info: *info}, nil
 }
 
+// Import represents the data about a mage:import package
 type Import struct {
 	Alias      string
 	Name       string
@@ -653,7 +665,7 @@ func getFunction(exp ast.Expr, pi *PkgInfo) (*Function, error) {
 	return nil, fmt.Errorf("unknown package for function %q", exp)
 }
 
-// getPackage returns the non-test package at the given path.
+// getPackage returns the importable package at the given path.
 func getPackage(path string, files []string, fset *token.FileSet) (*ast.Package, error) {
 	var filter func(f os.FileInfo) bool
 	if len(files) > 0 {
@@ -672,12 +684,21 @@ func getPackage(path string, files []string, fset *token.FileSet) (*ast.Package,
 		return nil, fmt.Errorf("failed to parse directory: %v", err)
 	}
 
-	for name, pkg := range pkgs {
-		if !strings.HasSuffix(name, "_test") {
-			return pkg, nil
+	switch len(pkgs) {
+	case 1:
+		var pkg *ast.Package
+		for _, pkg = range pkgs {
 		}
+		return pkg, nil
+	case 0:
+		return nil, fmt.Errorf("no importable packages found in %s", path)
+	default:
+		var names []string
+		for name := range pkgs {
+			names = append(names, name)
+		}
+		return nil, fmt.Errorf("multiple packages found in %s: %v", path, strings.Join(names, ", "))
 	}
-	return nil, fmt.Errorf("no non-test packages found in %s", path)
 }
 
 func hasContextParam(ft *ast.FuncType) bool {

--- a/parse/parse_test.go
+++ b/parse/parse_test.go
@@ -1,9 +1,17 @@
 package parse
 
 import (
+	"log"
+	"os"
 	"reflect"
 	"testing"
+
+	"github.com/magefile/mage/internal"
 )
+
+func init() {
+	internal.SetDebug(log.New(os.Stdout, "", 0))
+}
 
 func TestParse(t *testing.T) {
 	info, err := PrimaryPackage("go", "./testdata", []string{"func.go", "command.go", "alias.go", "repeating_synopsis.go", "subcommands.go"})
@@ -92,5 +100,15 @@ func TestParse(t *testing.T) {
 		if !found {
 			t.Fatalf("expected:\n%#v\n\nto be in:\n%#v", fn, info.Funcs)
 		}
+	}
+}
+
+func TestGetImportSelf(t *testing.T) {
+	imp, err := getImport("go", "github.com/magefile/mage/parse/testdata/importself", "")
+	if err != nil {
+		t.Fatal(err)
+	}
+	if imp.Info.AstPkg.Name != "importself" {
+		t.Fatalf("expected package importself, got %v", imp.Info.AstPkg.Name)
 	}
 }

--- a/parse/testdata/importself/magefile.go
+++ b/parse/testdata/importself/magefile.go
@@ -1,0 +1,14 @@
+//+build mage
+
+package main
+
+import (
+	"fmt"
+
+	// mage:import
+	_ "github.com/magefile/mage/parse/testdata/importself"
+)
+
+func Build(){
+	fmt.Println("built")
+}

--- a/parse/testdata/importself/targets.go
+++ b/parse/testdata/importself/targets.go
@@ -1,0 +1,7 @@
+package importself
+
+import "fmt"
+
+func Stuff() {
+	fmt.Println("stuff")
+}

